### ISSUE TITLE
Add utility functions: useFocusAtom, useSelectValue

### ISSUE THREAD

**Issue by KevinMusgrave** [pmndrs/jotai#721]
Instead of wrapping ```focusAtom``` with ```useAtom```, and ```selectAtom``` with ```useAtomValue```, I think the following would be nice:

### useFocusAtom
```javascript
import { useFocusAtom } from "jotai/optics";

function SomeComponent(props) {
  const [z, setZ] = useFocusAtom(x, (y) => y.prop("someKey"))
  return <></>
}
```

### useSelectValue
```javascript
import { useSelectValue } from "jotai/utils";

function SomeComponent(props) {
  const z = useSelectValue(x, (y) => y.someKey))
  return <></>
}
```

I guess the implementations would be something like this:
```javascript
export function useFocusAtom(anAtom, keyFn) {
  return useAtom(focusAtom(anAtom, keyFn))
}

export function useSelectValue(anAtom, keyFn) {
  return useAtomValue(selectValue(anAtom, keyFn))
}
```

**Comment by dai-shi** [comment]
We would like to export more atom-based utils and not hook-based utils unless they are technically possible (like depending on `useEffect`). We are moving towards it, and even deprecating some current hook-based utils is possible.

So, for focusAtom and selectAtom, we don't plan to export hooks. It's just few lines to create custom hooks on user's end as you showed.

---
Note that focusAtom and many others require stable callbacks.

This is what I would do:
```js
const useFocusAtom = (anAtom, fn) => useAtom(focusAtom(anAtom, fn))
```

But the usage requires useCallback:
```jsx
function SomeComponent(props) {
  const [z, setZ] = useFocusAtom(x, useCallback((y) => y.prop("someKey"), []))
  return <></>
}
```

If you prefer avoiding useCallback, you can do:
```js
const useFocusAtom = (anAtom, fn) => useAtom(focusAtom(anAtom, useRef(fn).current))
```

But, this hook doesn't support updating fn based on props.
So, it's really up to library users how to create hooks.

That said, I think we shouldn't include these hooks in the library. What would be nice is to have a recipe in docs to describe several patterns like the ones above.

**Comment by LuckyPigeon** [comment]
After my research, I think we can add these hook as doc recipe, for each recipe we can add an example:

  * `atomWithStorage => useAtom(atomWithStorage(name, initValue)`
  * `atomWithObservable => useAtom(atomWithObservable(fn))`
  * `atomWithReset => [...useAtom(anResetAtom), useResetAtom(anResetAtom)]`
  * `atomWithDefault => [...useAtom(anDefaultAtom), useResetAtom(anDefaultAtom)]`
  * `atomWithHash => useAtom(atomWithHash(hashName, hash)`
  * `selectAtom => useAtom(selectAtom(anAtom, fn)[0]`
  * `splitAtom => useAtom(splitAtom(anAtom))`

**Comment by dai-shi** [comment]
Sounds good! Please go ahead and add a new file in advanced-recipes. You want to create a codesandbox example for each use case.

**Comment by LuckyPigeon** [comment]
Above are all for `advanced-recipes/utils`, and for integrations like `optics (focusAtom)`, `immer` ... etc. I think we can use another folder `advanced-recipes/integrations`.

**Comment by dai-shi** [comment]
@LuckyPigeon I would like to have them in a single file. Like `advanced-recipes/custom-useatom-hooks.mdx`, and have sections for "utils" and "integrations".

**Comment by LuckyPigeon** [comment]
@dai-shi 
How about we show the way of creating these utilities to users and link the use case to original page. For example, we'll have `useAtomWithStorage` utility, after showing the way of creating it, we'll link user back to `atomWithStorage` page, and they can watch the use case. 
This approach can let us spend more time on improving our example on the original pages.

**Comment by LuckyPigeon** [comment]
Simple preview looks like this.
![image](https://user-images.githubusercontent.com/32315294/147551628-f7bc440f-3445-4464-96be-380dea0a5b2a.png)

**Comment by dai-shi** [comment]
Ah, I just found they don't work. You need `useMemo()` for atom creation.
We might want to focus on `useFocusAtom` and `useSelectAtom`.
I think you first want to create some examples with codesandbox.

I actually don't know the valid use case of `useAtomWithStorage` or so.

**Comment by LuckyPigeon** [comment]
I think `useAtomWithStorage` and other utility functions are creating for time saving.
If we want to only focus on `useFocusAtom` and `useSelectAtom`, I can make some changes and give these two an example.

**Comment by dai-shi** [comment]
`useAtom(atomWithStorage(...))` is basically `useAtom(atom(...))` which is not very intended because it's the same as `useState`.
If you don't use the atom elsewhere `useState` is a better.

Same for `atomWithStroage()` if the created atom is not used elsewhere, useState+useEffect is better.

Hm, so, I'd say useAtomWithStorage is a bad practice.

focusAtom and selectAtom make sense because we **have** base atoms elsewhere.

**Comment by LuckyPigeon** [comment]
OK. I think I finally got the idea. Thanks for the explanation @dai-shi. According to this rule, I think we can also add `freezeAtom` and `splitAtom` into advance-recipes, these two also have base atoms elsewhere.

**Comment by dai-shi** [comment]
Exactly!

**Comment by dai-shi** [comment]
closed by #923
